subsys/bluetooth/host/gatt.c

```diff
@@ -46,18 +46,48 @@ struct ccc_store {
 static sys_slist_t subscriptions;
 #endif /* CONFIG_BT_GATT_CLIENT */
 
-static const char *gap_name = CONFIG_BT_DEVICE_NAME;
 static const u16_t gap_appearance = CONFIG_BT_DEVICE_APPEARANCE;
 
 static sys_slist_t db;
 
 static ssize_t read_name(struct bt_conn *conn, const struct bt_gatt_attr *attr,
 			 void *buf, u16_t len, u16_t offset)
 {
-	return bt_gatt_attr_read(conn, attr, buf, len, offset, gap_name,
-				 strlen(gap_name));
+	const char *name = attr->user_data;
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, name,
+				 strlen(name));
+}
+
+#if CONFIG_BT_DEVICE_NAME_MAX > 0
+
+static ssize_t write_name(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			 const void *buf, u16_t len, u16_t offset,
+			 u8_t flags)
+{
+	char value[CONFIG_BT_DEVICE_NAME_MAX] = {};
+
+	if (flags & BT_GATT_WRITE_FLAG_PREPARE) {
+		return len;
+	}
+
+	if (offset > sizeof(value)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
```

Comment:
Wouldn't it make sense to return this error for any non-zero offset? If you copy to a non-zero offset that you end up with an empty string (since the first byte will be zero), which makes little sense and would likely anyway be different from what the GATT client expects. It would also simplify the code below.

Reply:
Yep, seems this should behave as a string it makes sense to not allow non-zero offset.